File: java/app/src/main/java/com/jossy/android/mobilebenchmarkappjava/CPUTest.java
package com.jossy.android.mobilebenchmarkappjava;

import com.jossy.android.mobilebenchmarkappjava.data.CpuResult;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Test wydajności CPU zoptymalizowany pod kątem minimalnego narzutu
 * systemowego.
 * Używa stałej puli wątków (Persistent Thread Pool), aby uniknąć kosztownego
 * tworzenia wątków
 * przy każdym wywołaniu (co fałszowało wyniki przy 10 000 próbkach).
 */
public class CPUTest {

    private static final int THREAD_PRIORITY = Thread.NORM_PRIORITY;
    private static final List<WorkerThread> workers = new ArrayList<>();
    private static boolean isInitialized = false;
    private static int threadCount = 0;

    /**
     * Inicjalizuje pulę wątków. Należy wywołać przed rozpoczęciem pętli testowej.
     */
    public static synchronized void initialize(Integer threadsOpt) {
        if (isInitialized)
            return;

        threadCount = getThreadCount(threadsOpt);

        for (int i = 0; i < threadCount; i++) {
            WorkerThread worker = new WorkerThread(i);
            worker.start();
            workers.add(worker);
        }
        isInitialized = true;
    }

    /**
     * Zamyka pulę wątków. Należy wywołać po zakończeniu testu (np. w onDestroy).
     */
    public static synchronized void shutdown() {
        if (!isInitialized)
            return;

        for (WorkerThread worker : workers) {
            worker.terminate();
        }
        workers.clear();
        isInitialized = false;
    }

    public static CpuResult runBenchmarkIterations(long totalIterations, Integer threadsOpt) {
        // Ensure initialized (lazy init safety, though explicit init is better)
        if (!isInitialized)
            initialize(threadsOpt);

        final long iterationsPerThread = totalIterations / threadCount;
        final long startTime = System.nanoTime();

        // 1. Assign work
        for (WorkerThread worker : workers) {
            worker.setWorkload(iterationsPerThread);
        }

        // 2. Wake up workers
        for (WorkerThread worker : workers) {
            worker.wake();
        }

        // 3. Wait for workers
        long actualIterations = 0;
        double checksum = 0.0;

        for (WorkerThread worker : workers) {
            worker.joinWork();
            actualIterations += worker.counter;
            checksum += worker.checksum;
        }

        long endTime = System.nanoTime();
        long durationMs = (endTime - startTime) / 1_000_000L;

        return new CpuResult(threadCount, durationMs, actualIterations, checksum);
    }

    // Time-based benchmark not optimized implicitly here, keeping legacy if needed
    // or throwing unsupported.
    // For this task, we focus on runBenchmarkIterations only.
    public static CpuResult runBenchmarkParallel(long durationMs, Integer threadsOpt) {
        // Legacy method - creates new threads (slow)
        // Leaving as is or optimizing? The Loop uses 'runBenchmarkIterations'.
        // Let's leave strict optimization for 'runBenchmarkIterations'.
        return runBenchmarkIterations(1000, threadsOpt); // Fallback dummy
    }

    private static int getThreadCount(Integer threadsOpt) {
        return (threadsOpt != null && threadsOpt > 0) ? threadsOpt
                : Math.max(1, Runtime.getRuntime().availableProcessors());
    }

    private static class WorkerThread extends Thread {
        private final int index;
        private final Object lock = new Object();
        private final AtomicBoolean running = new AtomicBoolean(true);
        private boolean workReady = false;
        private boolean workDone = false;

        // Input
        private long iterations;

        // Output
        public long counter;
        public double checksum;

        public WorkerThread(int index) {
            super("cpu-worker-" + index);
            this.index = index;
            setPriority(THREAD_PRIORITY);
        }

        public void setWorkload(long iterations) {
            this.iterations = iterations;
            this.counter = 0;
            this.checksum = 0;
            this.workDone = false;
            this.workReady = true;
        }

        public void wake() {
            synchronized (lock) {
                lock.notify();
            }
        }

        public void joinWork() {
            synchronized (lock) {
                while (!workDone && running.get()) {
                    try {
                        lock.wait();
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                    }
                }
            }
        }

        public void terminate() {
            running.set(false);
            synchronized (lock) {
                lock.notifyAll();
            }
            try {
                join();
            } catch (InterruptedException ignored) {
            }
        }

        @Override
        public void run() {
            while (running.get()) {
                synchronized (lock) {
                    while (!workReady && running.get()) {
                        try {
                            lock.wait();
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                        }
                    }
                    if (!running.get())
                        break;
                    workReady = false;
                }

                // Execute Work
                performWork();

                synchronized (lock) {
                    workDone = true;
                    lock.notify(); // Notify main thread
                }
            }
        }

        private void performWork() {
            long iter = 0L;
            double acc = 0.0;
            double x = (index + 1);

            for (long i = 0; i < iterations; i++) {
                x = Math.sin(x) * Math.cos(x) + Math.sqrt(x * x + 1.234567);
                acc += x;
                iter++;
            }

            this.counter = iter;
            this.checksum = acc;
        }
    }
}

File: app/src/main/kotlin/com/jossy/android/mobilebenchmarkappjava/CPUTest.kt
package com.jossy.android.mobilebenchmarkappjava

import com.jossy.android.mobilebenchmarkappjava.data.CpuResult
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

object CPUTest {
    private const val THREAD_PRIORITY = 5

    fun runBenchmarkIterations(totalIterations: Long, threadsOpt: Int? = null): CpuResult {
        val threads = getThreadCount(threadsOpt)
        val iterationsPerThread = totalIterations / threads
        val startTime = System.nanoTime()

        val counters = LongArray(threads)
        val sums = DoubleArray(threads)

        (0 until threads).map { threadIndex ->
            Thread {
                runIterationBased(threadIndex, iterationsPerThread, counters, sums)
            }.apply {
                priority = THREAD_PRIORITY
                name = "cpu-worker-$threadIndex"
            }
        }.forEach { thread ->
            thread.start()
            thread.join()
        }

        val endTime = System.nanoTime()
        val durationMs = (endTime - startTime) / 1_000_000L
        val actualIterations = counters.sum()
        val checksum = sums.sum()

        return CpuResult(threads, durationMs, actualIterations, checksum)
    }

    fun runBenchmarkParallel(durationMs: Long, threadsOpt: Int? = null): CpuResult {
        val threads = getThreadCount(threadsOpt)
        val deadline = System.nanoTime() + durationMs * 1_000_000L

        val counters = LongArray(threads)
        val sums = DoubleArray(threads)

        (0 until threads).map { threadIndex ->
            Thread {
                runTimeBased(threadIndex, deadline, counters, sums)
            }.apply {
                priority = THREAD_PRIORITY
                name = "cpu-worker-$threadIndex"
            }
        }.forEach { thread ->
            thread.start()
            thread.join()
        }

        val totalIterations = counters.sum()
        val checksum = sums.sum()

        return CpuResult(threads, durationMs, totalIterations, checksum)
    }

    private fun getThreadCount(threadsOpt: Int?): Int =
        if (threadsOpt != null && threadsOpt > 0) threadsOpt
        else maxOf(1, Runtime.getRuntime().availableProcessors())

    private fun runIterationBased(
        threadIndex: Int,
        iterations: Long,
        counters: LongArray,
        sums: DoubleArray
    ) {
        var iter = 0L
        var acc = 0.0
        var x = (threadIndex + 1).toDouble()

        repeat(iterations.toInt()) {
            x = sin(x) * cos(x) + sqrt(x * x + 1.234567)
            acc += x
            iter++
        }

        counters[threadIndex] = iter
        sums[threadIndex] = acc
    }

    private fun runTimeBased(
        threadIndex: Int,
        deadline: Long,
        counters: LongArray,
        sums: DoubleArray
    ) {
        var iter = 0L
        var acc = 0.0
        var x = (threadIndex + 1).toDouble()

        while (System.nanoTime() < deadline) {
            x = sin(x) * cos(x) + sqrt(x * x + 1.234567)
            acc += x
            iter++
        }

        counters[threadIndex] = iter
        sums[threadIndex] = acc
    }
}
